### PR TITLE
Highway To Ferry Tag Check

### PR DESCRIPTION
**Pull request by sayas01**
### Description:

This PR is to add a new Integrity Check - HighwayToFerryTagCheck. The purpose of this check is to flag all Edges with *route=FERRY* and *highway=PATH* (or higher). According to [OSM wiki](https://wiki.openstreetmap.org/wiki/Key:ferry), when *route=FERRY* exists, the ferry Tag can take values of highway classification and so, it is recommended to have a ferry Tag with the highway Tag value instead of the highway Tag itself. 

### Potential Impact:

Flags Edges that are Ferry routes if there is no corresponding Ferry tag with a highway classification or if it has both Ferry and Highway tags.

### Unit Test Approach:

Various unit tests have been added to test the functionality of the check.

### Test Results:

All tests passing successfully.

### CODE REVIEW

File: config/configuration.json [hunk 1]
```diff
@@ -413,5 +413,9 @@
       "difficulty": "NORMAL",
       "defaultPriority": "MEDIUM"
     }
+  },
+  "HighwayToFerryTagCheck": {
+    "enabled": true,
```
Comment (Bentleysb) [inline config/configuration.json]: This should not be included, as it can be obnoxious when trying to test other checks.

**Review by Bentleysb — APPROVED** [review]
LGTM. Just one little thing.

File: config/configuration.json (hunk 1 above)
Comment (sayas01) [reply config/configuration.json]: It was added when testing. Have removed it.

File: config/configuration.json [hunk 2]
```diff
@@ -413,5 +413,8 @@
       "difficulty": "NORMAL",
       "defaultPriority": "MEDIUM"
     }
+  },
+  "HighwayToFerryTagCheck": {
+    "highway.type.minimum": "path"
```
Comment (nachtm) [inline config/configuration.json]: Could we get MapRoulette instructions?

File: src/main/java/org/openstreetmap/atlas/checks/validation/tag/HighwayToFerryTagCheck.java [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+     * @param object
+     *            {@link AtlasObject} whose {@link HighwayTag} is verified for minimum highway value
+     * @return true if the highway tag of the object is greater or equal to the minimum highway type
+     */
+    private boolean isMinimumHighwayType(final AtlasObject object)
+    {
+        final Optional<HighwayTag> highwayTagOfObject = HighwayTag.highwayTag(object);
+
+        return highwayTagOfObject.isPresent()
+                && highwayTagOfObject.get().isMoreImportantThanOrEqualTo(this.minimumHighwayType);
```
Comment (nachtm) [inline src/main/java/org/openstreetmap/atlas/checks/validation/tag/HighwayToFerryTagCheck.java]: nit: We could get a one-liner here: 

`HighwayTag.highwayTag(object).map(tag -> tag.isMoreImportantThanOrEqualTo(this.minimumHighwayType)).orElse(false)`

File: src/main/java/org/openstreetmap/atlas/checks/validation/tag/HighwayToFerryTagCheck.java [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+     *
+     * @param configuration
+     *            {@link Configuration} required to construct any Check
+     */
+    public HighwayToFerryTagCheck(final Configuration configuration)
+    {
+        super(configuration);
+        final String highwayType = (String) this.configurationValue(configuration,
+                "highway.type.minimum", MINIMUM_HIGHWAY_TYPE_DEFAULT);
+        this.minimumHighwayType = Enum.valueOf(HighwayTag.class, highwayType.toUpperCase());
```
Comment (nachtm) [inline src/main/java/org/openstreetmap/atlas/checks/validation/tag/HighwayToFerryTagCheck.java]: Could you use the other `configurationValue` function above and put this line (converting from string to highway tag) as a lambda as the transform parameter?

File: src/main/java/org/openstreetmap/atlas/checks/validation/tag/HighwayToFerryTagCheck.java [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+                && Validators.isOfType(object, RouteTag.class, RouteTag.FERRY)
+                && this.isMinimumHighwayType(object) && !this.isFlagged(object.getOsmIdentifier());
+    }
+
+    @Override
+    protected Optional<CheckFlag> flag(final AtlasObject object)
+    {
+        // Mark OSM id as flagged
+        this.markAsFlagged(object.getOsmIdentifier());
+        final boolean hasHighwayClassification = this.hasSameClassificationAsHighwayTag(object);
```
Comment (nachtm) [inline src/main/java/org/openstreetmap/atlas/checks/validation/tag/HighwayToFerryTagCheck.java]: nit: hasSameHighwayClassification?

**Review by nachtm — COMMENTED** [review]
Nicely done @sayas01!! A couple comments:

**Review by nachtm — APPROVED** [review]
Nice fixes :)

**Review by danielduhh — CHANGES_REQUESTED** [review]
Hey Sayana; nice job with the check. You're missing one important document that we need for open source

**Comment by sayas01** [comment]
@danielduhh I have added the document for the check.

**Review by danielduhh — APPROVED** [review]
Thank you !  👊🏿

**Pull request closed without merging**